lib/util/query.js

```diff
@@ -1,130 +1,169 @@
+const { Readable } = require('stream');
+const _            = require('lodash');
+const promisify    = require('./promisify');
+
+const errors = {
+  DB_PARAM_REQUIRED: '"db" parameter is required',
+  MISSING_QUERY_DIR: '"queryDirectory" driver option must be set in order to use #namedQuery()',
+  MISSING_CQL: 'either #query() or #namedQuery() should be called prior to #execute()',
+  UNKNOWN: 'Error executing query'
+};
 
+class Query {
+  constructor(db, context) {
+    if (!db) {
+      throw new Error(errors.DB_PARAM_REQUIRED);
+    }
+
+    this.db = db;
+    this.context = _.assign({}, {
+      cql: null,
+      params: [],
+      options: {},
+      errors: [],
+      single: false,
+      first: false,
+      resultTransformers: []
+    }, context);
+  }
 
-var _         = require('lodash'),
-  through   = require('through2'),
-  readonly  = require('read-only-stream'),
-  isStream  = require('isstream'),
-  promisify = require('./promisify');
+  single() {
+    this.context.single = true;
+    this.context.first = false;
+    return this;
+  }
 
-function Query(db, context) {
-  if (!db) {
-    throw new Error('"db" parameter is required');
+  first() {
+    this.context.first = true;
+    this.context.single = false;
+    return this;
   }
 
-  this.db = db;
-  this.context = _.assign({}, {
-    cql: null,
-    params: [],
-    options: {},
-    errors: [],
-    single: false,
-    first: false,
-    resultTransformers: []
-  }, context);
-}
+  all() {
+    this.context.single = false;
+    this.context.first = false;
+    return this;
+  }
 
-Query.prototype.single = function single() {
-  this.context.single = true;
-  this.context.first = false;
-  return this;
-};
+  query(cql) {
+    this.context.cql = cql;
+    return this;
+  }
 
-Query.prototype.first = function single() {
-  this.context.first = true;
-  this.context.single = false;
-  return this;
-};
+  namedQuery(queryName) {
+    var self = this;
+
+    if (!self.db.queryCache) {
+      self.context.errors.push(new Error(errors.MISSING_QUERY_DIR));
+    } else {
+      self.db.queryCache.readQuery(
+        queryName,
+        function (err, queryText) {
+          if (err) {
+            return void self.context.errors.push(err);
+          }
 
-Query.prototype.all = function all() {
-  this.context.single = false;
-  this.context.first = false;
-  return this;
-};
+          addIfNotExists(self.context.options, 'queryName', queryName);
+          if (self.db.poolConfig.supportsPreparedStatements) {
```

Comment:
This `supportsPreparedStatements` is also an anachronysm. The `helenus` driver did not support prepared statements - everything had to be sent in text. With the new driver, always default to using prepared statements.

Reply:
@scommisso: done